firmware/shared/src/hw/hw_runTimeStat.c

```diff
@@ -0,0 +1,145 @@
+// Needed to ensure we are not running this does not compile for the bootloadeer binary
+
+#ifndef BOOTLOADER
+#include "hw_runTimeStat.h"
+#include "main.h"
+#include <string.h>
+#include <cmsis_os.h>
+#include "app_utils.h"
+#include "tasks.h"
+#include "util_errorCodes.h"
+
+/*
+ * Software defines
+ */
+#define IDLE_TASK_INDEX 0U
+#define TMR_SVC_INDEX 1U
+#define NUM_FT_TASKS 2U
+#define NUM_TOTAL_TASKS (NUM_FT_TASKS + NUM_OF_TASKS)
+
+/*
+ * Global Variables
+ */
+static TIM_HandleTypeDef *runTimeCounter;
+
+volatile unsigned long ulHighFrequencyTimerTick = 0;
+
+static TaskRuntimeStats *tasks_runtime_stat
+    [NUM_TOTAL_TASKS] = { [IDLE_TASK_INDEX] = &(TaskRuntimeStats){ 0 }, [TMR_SVC_INDEX] = &(TaskRuntimeStats){ 0 } };
+// We are going to assume that from here forth all our processors are going
+// single core but if not make the following an array
+
+static CpuRunTimeStats *cpu_runtime_stat;
+
+/*
+ * Function Decleration
+ */
+
+ExitCode hw_runTimeStat_init(TIM_HandleTypeDef *htim)
+{
+    if (htim == NULL)
+    {
+        return EXIT_CODE_INVALID_ARGS;
+    }
+
+    runTimeCounter = htim;
+
+    return EXIT_CODE_OK;
+}
+
+ExitCode hw_runtimeStat_registerCpu(CpuRunTimeStats *cpu_info)
+{
+    if (cpu_info->cpu_usage_max_setter == NULL || cpu_info->cpu_usage_setter == NULL)
+    {
+        return EXIT_CODE_INVALID_ARGS;
+    }
+
+    cpu_runtime_stat = cpu_info;
+
+    return EXIT_CODE_OK;
+}
+
+ExitCode hw_runTimeStat_registerTask(TaskRuntimeStats *task_info)
+{
+    if (task_info->cpu_usage_max_setter == NULL || task_info->cpu_usage_setter == NULL ||
+        task_info->stack_usage_max_setter == NULL || task_info->stack_size != 0)
+    {
+        return EXIT_CODE_INVALID_ARGS;
+    }
+    else if (NUM_OF_TASKS < task_info->task_index)
+    {
+        return EXIT_CODE_OUT_OF_RANGE;
+    }
+
+    tasks_runtime_stat[task_info->task_index + NUM_FT_TASKS] = task_info;
```

Comment:
any particular reason we need to use task_index for this? I am not a big fan of the inverted dependency situation where you need to circular import